### PR TITLE
Feat/al2023 support

### PR DESCRIPTION
**Pull request by dm1tr1yvovk**

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -1,21 +1,125 @@
...
+  verify-dist:
+    name: Verify dist is up to date
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
```
Comment (dm1tr1yvovk) [inline .github/workflows/pr.yml]: update to 24 here?

**Review by dm1tr1yvovk — COMMENTED** [review]
# Code Review: PR #28 `feat/al2023-support`

## Strengths

1. **Clean AWS SDK v3 migration** — Import-only-what-you-need command pattern (`client.send(new XCommand(...))`) is correct and reduces bundle size. (`src/aws.js:1-8`)

2. **Robust security hardening** — IMDSv2 enforcement with `HttpPutResponseHopLimit: 1` is the correct SSRF defense. Default-on with `http-tokens: optional` escape hatch is a good backward-compat trade-off. (`src/aws.js:201-210`)

3. **Non-root runner + ephemeral mode** — Running as a dedicated `runner` user with `--ephemeral --disableupdate` is a significant security win. The heredoc-quoted `<<'RUNNER_BOOTSTRAP'` correctly prevents shell expansion of untrusted values. (`src/aws.js:138-186`)

4. **Checksum verification is enforced, not advisory** — `sha256sum -c -` runs under `set -euo pipefail`, killing bootstrap on mismatch. The CI job `verify-runner-url` cross-checks the hardcoded table against upstream releases. Two-layer defense, well designed. (`src/aws.js:177`, `.github/workflows/pr.yml:92-125`)

5. **Independent cleanup in stop mode** — Catching errors from `terminateEc2Instance()` and `removeRunner()` independently prevents a GitHub API outage from blocking EC2 termination (and thus preventing billing runaway). (`src/index.js:54-84`)

6. **Structured logging with sanitization** — The `log.js` module with key-based secret redaction and deferred `require('./config')` inside `emit()` correctly handles circular dependency with module load order. (`src/log.js:28-41`)

7. **Comprehensive test coverage** — All 52 tests pass. Covers config validation, EBS mapping edge cases (6 scenarios), retry backoff, checksum integrity, and log sanitization.

8. **CI pipeline additions** — `verify-dist`, `unit-tests`, and `verify-runner-url` jobs catch real defect classes at review time.

---

## Issues

### Critical (Must Fix)

**C1. `_.merge()` mutates `config.githubContext` — stale `runner_id` on subsequent calls**
- **File:** `src/gh.js:55`
- **What's wrong:** `_.merge(config.githubContext, { runner_id: runner.id })` mutates its first argument in place. After `removeRunner()` runs, `config.githubContext` permanently gains a stale `runner_id` property, which will contaminate any subsequent logging or API calls using that object.
- **Fix:** Use non-mutating merge: `{ ...config.githubContext, runner_id: runner.id }` or `_.merge({}, config.githubContext, { runner_id: runner.id })`

**C2. Missing `ec2:DescribeImages` IAM permission in documented policy**
- **File:** `README.md:131-143`
- **What's wrong:** `resolveImage()` now always calls `DescribeImagesCommand` even when `ec2-image-id` is provided directly (to fetch BlockDeviceMappings for EBS encryption). But `ec2:DescribeImages` is not in the documented minimum IAM policy. **Users upgrading will get `AccessDenied` errors.**
- **Fix:** Add `ec2:DescribeImages` to the README's IAM policy block.

---

### Important (Should Fix)

**I1. Stale DEP0169 suppression after SDK v3 migration**
- **File:** `src/index.js:1-14`
- **What's wrong:** The `url.parse()` deprecation suppression was added for aws-sdk v2. The comment says "migrating to aws-sdk v3 would require rewriting the entire action" — but that migration is already done. Dead code with a misleading comment.
- **Fix:** Remove lines 1-14 entirely.

**I2. `waitForRunnerRegistered` race — both reject and resolve can fire**
- **File:** `src/gh.js:82-93`
- **What's wrong:** When timeout fires, `reject()` on line 86 has no `return`, so execution falls through to line 89 where `resolve()` may also fire if the runner registered at the exact timeout boundary.
- **Fix:** Add `return;` after `reject(...)` on line 86.

**I3. New inputs not documented in README inputs table**
- **File:** `README.md:269-282`
- **What's wrong:** Four new inputs (`runner-version`, `encrypt-ebs`, `http-tokens`, `debug`) are in `action.yml` but absent from the README's Inputs table. Users won't discover the new security features.

**I4. README still references "Amazon Linux 2" instead of AL2023**
- **File:** `README.md:275`
- **What's wrong:** Says "compatible with Amazon Linux 2 images" — contradicts the entire purpose of this PR.

**I5. No validation of `http-tokens` input value**
- **File:** `src/config.js:20`
- **What's wrong:** Passed directly to `MetadataOptions.HttpTokens` without validation. A typo like `require` instead of `required` produces a confusing AWS API error at instance launch rather than a clear validation error at action start.
- **Fix:** `if (!['required', 'optional'].includes(this.input.httpTokens)) throw new Error(...)`

**I6. `ec2Client()` creates a new client on every call despite "singleton" comment**
- **File:** `src/aws.js:16-22`
- **What's wrong:** Comment says "A single shared client is fine" but the function creates a new `EC2Client` per invocation. Wastes credential resolution cycles. Comment and behavior are contradictory.
- **Fix:** Either make it a true module-scope singleton or update the comment.

---

### Minor (Nice to Have)

**M1.** `src/gh.js:86` — `reject()` with a plain string instead of `Error` object. The error handler in `index.js:90` accesses `.name` and `.message`, which will be `undefined`.

**M2.** `package.json:21-22` — `@actions/core` at `^1.2.6` (2020) and `@actions/github` at `^4.0.0` (2021) are very outdated. Upgrading to `@actions/core` v1.10+ would eliminate the custom `setOutput()` function.

**M3.** `README.md:319,324-325` — Example workflow still uses `aws-actions/configure-aws-credentials@v1` and static IAM keys, contradicting the new OIDC-preferred guidance.

**M4.** `src/aws.js:212` — `encryptEbs === 'true'` string comparison works but `core.getBooleanInput()` would be more conventional.

**M5.** `src/aws.js:46-51` — Redundant `DescribeImages` call when `ec2-image-id` is provided and `encrypt-ebs` is false. Adds latency and forces the new permission even when not needed. Fixing this also reduces the impact of C2.

---

## Recommendations

1. **Add a manual integration test workflow** (`workflow_dispatch`) that creates/tears down a test runner to catch bootstrap failures, IAM permission gaps, and SDK issues that unit tests can't.

2. **Pin `@aws-sdk/client-ec2` with caret range** (`^3.1033.0` instead of exact `3.1033.0`) to receive patch/security fixes, or document the rationale for the exact pin.

3. **Document KMS permissions for `encrypt-ebs`** — With the default AWS-managed key, the role needs `kms:CreateGrant` and `kms:DescribeKey` on `alias/aws/ebs`.

4. **Consider removing lodash** — The only usages are `_.filter` and `_.merge`, both trivially replaceable with native JS. This would eliminate a transitive dependency.

---

## Assessment

**Ready to merge?** No — with fixes.

**Reasoning:** The implementation is architecturally sound and delivers significant security and reliability improvements. The two critical issues must be addressed: C1 (`_.merge` mutation) is a correctness bug, and C2 (missing IAM permission in docs) is a breaking change for existing users. The stale DEP0169 suppression (I1) and the timeout race condition (I2) should also be fixed before merge. Remaining items can go in follow-ups.
